qualification/general/test_control.py

```diff
@@ -0,0 +1,206 @@
+        with check:
+            assert elapsed < 1.0
+
+
+@pytest.fixture
+async def sensor_watcher(cbf: CBFRemoteControl) -> AsyncGenerator[aiokatcp.SensorWatcher, None]:
+    # aiokatcp doesn't currently handle adding watchers after the connection
+    # is already established; SensorWatcher is also somewhat expensive. So
+    # instead we create a separate connection for monitoring sensors.
+    secondary = aiokatcp.Client(*cbf.product_controller_endpoint)
+    sensor_watcher = aiokatcp.SensorWatcher(secondary)
+    secondary.add_sensor_watcher(sensor_watcher)
+    await sensor_watcher.synced.wait()
+    yield sensor_watcher
+    secondary.close()
+    await secondary.wait_closed()
+
+
+def check_timestamps(
+    name: str,
+    receiver: XBReceiver,
+    pdf_report: Reporter,
+    timestamps: list[int],
+):
+    assert timestamps, f"No {name} chunks received"
+    elapsed = (timestamps[-1] - timestamps[0]) / receiver.scale_factor_timestamp
+    pdf_report.detail(f"{name}: received data over {elapsed:.3f}s.")
+    with check:
+        assert elapsed >= 55.0, f"Less than 55s of data received for {name}"
+    # Trim the first 5 seconds, where we expect to see some losses
+    start = timestamps[0] + 5 * receiver.scale_factor_timestamp
+    start_pos = 0
+    while timestamps[start_pos] < start:
+        start_pos += 1
+    del timestamps[:start_pos]
+    expected = (timestamps[-1] - timestamps[0]) // receiver.timestamp_step + 1
+    missing = expected - len(timestamps)
+    pdf_report.detail(f"{name}: missed {missing} of {expected} chunks (ignoring first 5s).")
+    with check:
+        assert missing <= 1, f"{missing} of {expected} chunks missing for {name}"
+
+
+# TODO: once requirements spec is finalised, note which requirements
+# this corresponds to.
+async def test_control(
+    cbf: CBFRemoteControl,
+    receive_baseline_correlation_products: BaselineCorrelationProductsReceiver,
+    receive_tied_array_channelised_voltage: TiedArrayChannelisedVoltageReceiver,
+    pdf_report: Reporter,
+    sensor_watcher: aiokatcp.SensorWatcher,
+) -> None:
+    """Test that controlling a correlator doesn't cause data to be lost.
+
+    Verification method
+    -------------------
+    Verified by means of test. Issue requests to the configured streams as follows:
+
+    - antenna_channelised_voltage: ``?delays`` every 6 seconds, with 3 seconds lead time.
+    - tied_array_channelised_voltage: ``?beam-delays`` every 3.4 seconds.
+
+    Check that each request completes in at most 1 second. Receive data from
+    the streams over 60s and check that the received chunks span at least 55
+    seconds and have no more than one chunk lost after the first 5 seconds
+    (during which we expect losses as the receiver starts with a backlog).
+
+    Additionally, subscribe to all sensor updates to emulate CAM's load.
+    """
+    pdf_report.step("Subscribe to sensors.")
+    pdf_report.detail(f"Subscribed to {len(sensor_watcher.sensors)} sensors.")
+
```

Comment:
There's no actual code here, this is happening ... in the `sensor_watcher` fixture if I understand correctly?

I'd just leave a comment indicating that this is the case. Or potentially put these lines inside the fixture.

Reply:
I've moved the await out of the fixture, so that there is actually some code here.